test/load-generator/acme/directory.go

```diff
@@ -0,0 +1,208 @@
+	RevokeCertEndpoint Endpoint = "revokeCert"
+	// KeyChangeEndpoint is the directory key for the keyChange endpoint.
+	KeyChangeEndpoint Endpoint = "keyChange"
+)
+
+var (
+	// ErrEmptyDirectory is returned if New is provided and empty directory URL.
+	ErrEmptyDirectory = errors.New("directoryURL must not be empty")
+	// ErrInvalidDirectoryURL is returned if New is provided an invalid directory URL.
+	ErrInvalidDirectoryURL = errors.New("directoryURL is not a valid URL")
+	// ErrInvalidDirectoryHTTPCode is returned if New is provided a directory URL
+	// that returns something other than HTTP Status OK to a GET request.
+	ErrInvalidDirectoryHTTPCode = errors.New("GET request to directoryURL did not result in HTTP Status 200")
+	// ErrInvalidDirectoryJSON is returned if New is provided a directory URL
+	// that returns invalid JSON.
+	ErrInvalidDirectoryJSON = errors.New("GET request to directoryURL returned invalid JSON")
+
+	// RequiredEndpoints is a slice of Endpoint keys that must be present in the
+	// ACME server's directory. The load-generator uses each of these endpoints
+	// and expects to be able to find a URL for each in the server's directory
+	// resource.
+	RequiredEndpoints = []Endpoint{
+		NewNonceEndpoint, NewAccountEndpoint,
+		NewOrderEndpoint, RevokeCertEndpoint,
+	}
+)
+
+// Endpoint represents a string key used for looking up an endpoint URL in an ACME
+// server directory resource.
+//
+// E.g. NewOrderEndpoint -> "newOrder" -> "https://acme.example.com/acme/v1/new-order-plz"
+//
+// See "ACME Resource Types" registry - RFC 8555 Section 9.7.5.
+type Endpoint string
+
+// ErrMissingEndpoint is an error returned if NewDirectory is provided an ACME
+// server directory URL that is missing a key for a required endpoint in the
+// response JSON. See also RequiredEndpoints.
+type ErrMissingEndpoint struct {
+	endpoint Endpoint
+}
+
+// Error returns the error message for an ErrMissingEndpoint error.
+func (e ErrMissingEndpoint) Error() string {
+	return fmt.Sprintf(
+		"directoryURL JSON was missing required key for %q endpoint",
+		e.endpoint,
+	)
+}
+
+// ErrInvalidEndpointURL is an error returned if NewDirectory is provided an
+// ACME server directory URL that has an invalid URL for a required endpoint.
+// See also RequiredEndpoints.
+type ErrInvalidEndpointURL struct {
+	endpoint Endpoint
+	value    string
+}
+
+// Error returns the error message for an ErrInvalidEndpointURL error.
+func (e ErrInvalidEndpointURL) Error() string {
+	return fmt.Sprintf(
+		"directoryURL JSON had invalid URL value (%q) for %q endpoint",
+		e.value, e.endpoint)
+}
+
+// Directory is a type for holding URLs extracted from the ACME server's
+// Directory resource.
+//
+// See RFC 8555 Section 7.1.1 "Directory".
+//
+// Its public API is read-only and therefore it is safe for concurrent access.
+type Directory struct {
+	// endpointURLs is a map from endpoint name to
+	endpointURLs map[Endpoint]*url.URL
```

Comment:
Is the main reason for using `url.URL`s to check the URL formatting is proper at the directory parsing stage rather than the HTTP request stage?

Reply:
(I mainly ask because using `url.URL` feels like it's a bit of a pain in the ass)

Reply:
Yes that's the major reason. I'll switch the map to be `map[Endpoint]string`.

Reply:
Fixed in 0cc0b2b